pkg/wireguard/operator/operator.go

```diff
@@ -0,0 +1,360 @@
+}
+
+type Operator struct {
+	lock.RWMutex
+	restoring             bool
+	ciliumNodeUpdater     CiliumNodeUpdater
+	ipv4Alloc             *ipallocator.Range
+	ipv4ByNode            map[string]net.IP
+	ipv4AllocAfterRestore map[string]struct{} // by nodename
+	ipv6Alloc             *ipallocator.Range
+	ipv6ByNode            map[string]net.IP
+	ipv6AllocAfterRestore map[string]struct{} // by nodename
+}
+
+func NewOperator(subnetV4, subnetV6 *net.IPNet, ciliumNodeUpdater CiliumNodeUpdater) (*Operator, error) {
+	var (
+		err                                          error
+		ipv4Alloc, ipv6Alloc                         *ipallocator.Range
+		ipv4ByNode, ipv6ByNode                       map[string]net.IP
+		ipv4AllocAfterRestore, ipv6AllocAfterRestore map[string]struct{}
+	)
+
+	if option.Config.EnableIPv4 {
+		ipv4ByNode = make(map[string]net.IP)
+		ipv4AllocAfterRestore = make(map[string]struct{})
+		ipv4Alloc, err = ipallocator.NewCIDRRange(subnetV4)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if option.Config.EnableIPv6 {
+		ipv6ByNode = make(map[string]net.IP)
+		ipv6AllocAfterRestore = make(map[string]struct{})
+		ipv6Alloc, err = ipallocator.NewCIDRRange(subnetV6)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	m := &Operator{
+		restoring:             true,
+		ciliumNodeUpdater:     ciliumNodeUpdater,
+		ipv4Alloc:             ipv4Alloc,
+		ipv4ByNode:            ipv4ByNode,
+		ipv4AllocAfterRestore: ipv4AllocAfterRestore,
+		ipv6Alloc:             ipv6Alloc,
+		ipv6ByNode:            ipv6ByNode,
+		ipv6AllocAfterRestore: ipv6AllocAfterRestore,
+	}
+
+	return m, nil
+}
+
+func (o *Operator) AddNode(n *v2.CiliumNode) error {
+	o.Lock()
+	defer o.Unlock()
+
+	if option.Config.EnableIPv4 {
+		if err := o.allocateIP(n, IPv4); err != nil {
+			return err
+		}
+	}
+
+	if option.Config.EnableIPv6 {
+		if err := o.allocateIP(n, IPv6); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (o *Operator) UpdateNode(n *v2.CiliumNode) error {
+	o.Lock()
+	defer o.Unlock()
+
+	if option.Config.EnableIPv4 {
+		if err := o.allocateIP(n, IPv4); err != nil {
+			return err
+		}
+	}
+
+	if option.Config.EnableIPv6 {
+		if err := o.allocateIP(n, IPv6); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (o *Operator) DeleteNode(n *v2.CiliumNode) {
+	o.Lock()
+	defer o.Unlock()
+
+	nodeName := n.ObjectMeta.Name
+
+	if o.restoring {
+		log.WithField("nodeName", nodeName).Warn("Received node delete while restoring")
+	}
+
+	if option.Config.EnableIPv4 {
+		o.releaseIP(n, nodeName, IPv4)
+	}
+
+	if option.Config.EnableIPv6 {
+		o.releaseIP(n, nodeName, IPv6)
+	}
+}
+
+func (o *Operator) releaseIP(n *v2.CiliumNode, nodeName string, family Family) {
+	var (
+		ipAlloc  *ipallocator.Range
+		ipByNode map[string]net.IP
+	)
+
+	switch family {
+	case IPv4:
+		ipAlloc = o.ipv4Alloc
+		ipByNode = o.ipv4ByNode
+	case IPv6:
+		ipAlloc = o.ipv6Alloc
+		ipByNode = o.ipv6ByNode
+	default:
+		log.Warning("unsupported family")
```

Comment:
I assume the the user can manually modify the CiliumNode resource and make a typo here, so I think this warning should provide much more context.

Reply:
The family here is actually only ever set by the programmer. We decided to do a panic here, since passing the wrong familiy is always a programmer error (and not something the user can address).